### PR TITLE
Pipeshot crafting and shells

### PR DESCRIPTION
**Pull request by NightmareChamillian**
[balance]
## About the PR 
Adds custom shells and reworks pipeshot crafting to be more intuitive, as well as allow for different item costs. Makes it so you need 2 glass shards for glass shot instead of 1.

## Why's this needed? 
more intuitive and allows for future pipeshot recipies to consume more than 1 item

## Changelog

```changelog
(u)NightmarechaMillian
(+)Glass pipeshot now requires 2 shards instead of 1.
```

### CODE REVIEW

File: code/obj/item/assembly/misc_assemblies.dm [hunk 1]
```diff
@@ -880,24 +880,39 @@ obj/item/assembly/radio_horn/receive_signal()
...
-			qdel(W)
-			qdel(src)
-			user.put_in_hand_or_drop(shot)
+		if (istype(W, /obj/item/raw_material/shard))
+			if (!crafttype)
+				crafttype = 1
+				thingsneeded = 2
+			if(crafttype == 1)
+				src.craftwith(W, src, user)
+		if (istype(W, /obj/item/raw_material/scrap_metal))
```
Comment (blackrep) [inline code/obj/item/assembly/misc_assemblies.dm:896]: One thing I've realized playing a couple of rounds trying to craft this. Scrap metal is extremely rare to find. The only way to obtain it is to purposely blow up flooring & walls. Or find it random-spawned in prefabs / maintenance flooring.

File: code/obj/item/assembly/misc_assemblies.dm [hunk 2]
```diff
@@ -880,24 +880,39 @@ obj/item/assembly/radio_horn/receive_signal()
 //////////////////////////////////handmade shotgun shells//////////////////////////////////
 
 /obj/item/assembly/makeshiftshell
-
 	name = "filled pipe hulls"
 	desc = "Four open pipe shells, with propellant in them. You wonder what you could stuff into them."
 	icon_state = "Pipeshotrow"
+	var/thingsneeded = null //number of things still required to produce ammunition
+	var/crafttype = null //1 is glass, 2 is scrap, 3 is plasmaglass
```
Comment (blackrep) [inline code/obj/item/assembly/misc_assemblies.dm:887]: Outdated comment?

File: code/obj/item/assembly/misc_assemblies.dm (hunk 1 above)
Comment (NightmareChamillian) [reply code/obj/item/assembly/misc_assemblies.dm:896]: That's intentional, to skew it's use towards antags and more dire situations/later in the round. On high pop I've usually been able to find 1-2 bits of scrap around the station/in debris even without blowing anything up, which is more or less how obtainable I want it to be.

File: code/obj/item/assembly/misc_assemblies.dm (hunk 2 above)
Comment (NightmareChamillian) [reply code/obj/item/assembly/misc_assemblies.dm:887]: oops

File: code/obj/item/assembly/misc_assemblies.dm [hunk 3]
```diff
@@ -880,24 +880,39 @@ obj/item/assembly/radio_horn/receive_signal()
...
-			qdel(W)
-			qdel(src)
-			user.put_in_hand_or_drop(shot)
-
-		if(istype(W, /obj/item/raw_material/scrap_metal))
-			var/obj/item/ammo/bullets/pipeshot/scrap/shot = new /obj/item/ammo/bullets/pipeshot/scrap/(get_turf(src))
-			qdel(W)
-			qdel(src)
-			user.put_in_hand_or_drop(shot)
+		if (!crafttype)
```
Comment (blackrep) [inline code/obj/item/assembly/misc_assemblies.dm:890]: My suggestion to make this attackby cleaner. Make the recipe handler a datum, with a child for each recipe:
1. When the first item is added, assign the datum
2. Let the datum handle inserting items into the shell (the logic of whether or not to accept the item, and whether or not it's done)
a. Datum should have a path to the craft material (shard / scrap_metal)
b. Datum should have a number needed of the craft material
c. Datum should have a path to the result ammo
d. Datum should have a proc to handle arbritrary attackby to check if it's the craft material
e. The proc in d should return whether or not the craft is done
4. When the datum returns that the craft is done, request the datum return the finished product and spawn it at the src location.

Comment (NightmareChamillian) [reply code/obj/item/assembly/misc_assemblies.dm:890]: good idea.

**Comment by NightmareChamillian** [comment]
setting this as a draft until I datumize things

**Review by Tarmunora — APPROVED** [review]

**Comment by blackrep** [comment]
Way cleaner than before!

**Pull request merged**